### PR TITLE
Adding helpers to parse Bigtable create cluster operation.

### PR DESCRIPTION
**Pull request by dhermes**
A `CreateCluster` request response doesn't actually indicate
success or failure. Rather it returns a `Cluster` message with
the validated request parts inside and a `current_operation`
attached.

We implement `_process_operation` so that we can determine the
ID of that long-running operation (so it can be checked for
completion / success, if desired by the user). In addition
we seek to notify the user when the request began.

From the [service definition][1] we know that the `current_operation`
is a [long-running operation][2] and that:

>  The embedded operation's "metadata" field type is `CreateClusterMetadata`,
>  The embedded operation's "response" field type is `Cluster`, if successful.

The [`Operation` metadata][3] is of type [`Any`][4] (which uses a `type_url`
and raw bytes to provide **any** protobuf message type in a single
field, but still allow it to be parsed into it's true type after
the fact). So we expect `CreateCluster` responses to have long-running
operations with a type URL matching [`CreateClusterMetadata`][5].

As a result, we introduce a utility (`_parse_pb_any_to_native`) for
parsing an `Any` field into the native protobuf type specified by the
type URL. Since we know we need to handle `CreateClusterMetadata` values,
we add a default mapping (`_TYPE_URL_MAP`) from the corresponding type url
for that message type to the native Python type.

The `CreateClusterMetadata` type has `request_time` and
`finish_time` fields of type [`Timestamp`][6] so we also
add the `_pb_timestamp_to_datetime` helper for converting protobuf
messages into native Python `datetime.datetime` objects.

[1]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/bigtable/admin/cluster/v1/bigtable_cluster_service.proto#L64
[2]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/bigtable/admin/cluster/v1/bigtable_cluster_data.proto#L74
[3]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/longrunning/operations.proto#L82
[4]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/protobuf/any.proto#L58
[5]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/bigtable/admin/cluster/v1/bigtable_cluster_service_messages.proto#L83-L92
[6]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/protobuf/timestamp.proto#L78

### CODE REVIEW

**Comment by dhermes** [comment]
@tseaver Sorry this is kind of a doozy. I wrote a long description to try to give you some context and quick windows into what the actual data returned from the API is.

-----

This PR is doing something very simple: parse the information from the `CreateCluster` response. But it takes a bit of machinery for doing this.

File: gcloud/bigtable/cluster.py [hunk 1]
```diff
@@ -74,6 +85,73 @@ def _prepare_create_request(cluster):
...
+    :rtype: :class:`datetime.datetime`
+    :returns: A UTC datetime object converted from a protobuf timestamp.
+    """
+    return (
+        _EPOCH +
+        datetime.timedelta(
+            seconds=timestamp.seconds,
+            microseconds=(timestamp.nanos / 1000.0),
+        )
+    )
```
Comment (tseaver) [inline gcloud/bigtable/cluster.py]: Should this be in a `_helpers` module?

File: gcloud/bigtable/cluster.py [hunk 2]
```diff
@@ -74,6 +85,73 @@ def _prepare_create_request(cluster):
...
+    :rtype: object
+    :returns: The de-serialized object.
+    :raises: :class:`ValueError <exceptions.ValueError>` if the
+             ``expected_type`` does not match the ``type_url`` on the input.
+    """
+    if expected_type is not None and expected_type != any_val.type_url:
+        raise ValueError('Expected type: %s, Received: %s' % (
+            expected_type, any_val.type_url))
+    container_class = _TYPE_URL_MAP[any_val.type_url]
+    return container_class.FromString(any_val.value)
```
Comment (tseaver) [inline gcloud/bigtable/cluster.py]: Should this be in a `_helpers` module?

File: gcloud/bigtable/cluster.py [hunk 3]
```diff
@@ -74,6 +85,73 @@ def _prepare_create_request(cluster):
...
+    if match is None:
+        raise ValueError('Cluster create operation name was not in the '
+                         'expected format.', operation_pb.name)
+    operation_id = int(match.group('operation_id'))
+
+    request_metadata = _parse_pb_any_to_native(operation_pb.metadata)
+    operation_begin = _pb_timestamp_to_datetime(
+        request_metadata.request_time)
+
+    return operation_id, operation_begin
```
Comment (tseaver) [inline gcloud/bigtable/cluster.py]: Are we only going to be parsing "operations" pb's in this one place?  If so, maybe the function should have a less generic name (`_process_cluster_create_operation`?)  If not, maybe it should move to a `_helpers` module, and take the regex as an argument?

File: gcloud/bigtable/cluster.py (hunk 1 above)
Comment (dhermes) [inline gcloud/bigtable/cluster.py]: I've been going back on forth on that. Do you have a reason to put them one place or another? I can't decide.

-----

The https://github.com/dhermes/gcloud-python-bigtable is "done" and the only place these helpers ended up being used were `cluster.py`. They actually live in `gcloud_python._helpers` over there, so I moved them to have them closer to the place they are used. Their existence is just for "code quality" and making it easier to test things.

Comment (tseaver) [inline gcloud/bigtable/cluster.py]: Given that you know they won't be used elsewhere, I'm fine with leaving them here.

File: gcloud/bigtable/cluster.py (hunk 3 above)
Comment (dhermes) [inline gcloud/bigtable/cluster.py]: Yes of the three Bigtable services

- [Cluster][1]
- [Table][2]
- [Data][3]

only the cluster service uses long-running operations. It uses them for `Create`, `Update` and `Undelete`.

[1]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/bigtable/admin/cluster/v1/bigtable_cluster_service.proto#L31
[2]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/bigtable/admin/table/v1/bigtable_table_service.proto#L31
[3]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/8e363d72eb39d921dfdf5daf4a36032aa9d003e2/bigtable-protos/src/main/proto/google/bigtable/v1/bigtable_service.proto#L31

File: gcloud/bigtable/cluster.py (hunk 1 above)
Comment (dhermes) [inline gcloud/bigtable/cluster.py]: SGTM

File: gcloud/bigtable/cluster.py (hunk 3 above)
Comment (tseaver) [inline gcloud/bigtable/cluster.py]: Do the other uses of them end up calling `_process_operation`?  If so, the message for the `ValueError` is maybe misleading.

Comment (dhermes) [inline gcloud/bigtable/cluster.py]: Yes all three (`Create`, `Update` and `Undelete`) call `_process_operation`.

Such a stupid error message! Nice catch. I'll ping you once fixed.

File: gcloud/bigtable/cluster.py (hunk 2 above)
Comment (tseaver) [inline gcloud/bigtable/cluster.py]: As with above, fine to leave in place.

Comment (dhermes) [inline gcloud/bigtable/cluster.py]: Thanks

File: gcloud/bigtable/cluster.py (hunk 3 above)
Comment (dhermes) [inline gcloud/bigtable/cluster.py]: @tseaver PTAL. I updated the string in the `ValueError`. (Let's hope no user ever sees it, that would mean Google failed to send the correct operation ID format or changed it under our nose.)

**Comment by tseaver** [comment]
LGTM

**Pull request merged**